src/include/gcc.h

```diff
@@ -27,71 +27,145 @@
- * that never happens, we check address alignment: we know of no architectures
- * with cache lines other than a multiple of 4 bytes in size, so aligned 4-byte
- * accesses will always be in a single cache line.
+ * Violating this can lead to races when two threads are cooperating. Suppose there is a thread X
+ * that finds the variable A of 32-bits set to 0 and writes in a new value. Concurrently another
+ * thread Y reads the adjacent variable B of 32-bits that are non-zero, does some operation, resets
+ * the memory location to 0. If thread X were to read the 64-bits consisting of the variable A and
+ * the adjacent variable B together, and write them both, the two threads could race.
  *
- * Atomic writes are often associated with memory barriers, implemented by the
- * WT_ACQUIRE_BARRIER and WT_RELEASE_BARRIER macros.  WiredTiger's requirement as
- * described by the Solaris membar_enter description:
+ * # Writes Across Cache Lines #
+ * WiredTiger doesn't depend on writes across cache lines being atomic and it ensures that a single
+ * write that is smaller or equal to 4 bytes long is never split across the cache lines. This is
+ * achieved through address alignment: we know of no architectures with cache lines other than a
+ * multiple of 4 bytes in size, therefore, aligned 4-byte accesses will always be in a single cache
+ * line.
  *
- *	No stores from after the memory barrier will reach visibility and
- *	no loads from after the barrier will be resolved before the lock
- *	acquisition reaches global visibility
+ * # Cache coherency #
+ * WiredTiger depends on running on a coherent cache, it cannot run on hardware that doesn't
+ * implement a typical cache coherence protocol.
  *
- * In other words, the WT_RELEASE_BARRIER macro must ensure that memory stores by
- * the processor, made before the WT_RELEASE_BARRIER call, be visible to all
- * processors in the system before any memory stores by the processor, made
- * after the WT_RELEASE_BARRIER call, are visible to any processor.  The
- * WT_ACQUIRE_BARRIER macro ensures that all loads before the barrier are complete
- * before any loads after the barrier.  The compiler cannot reorder or cache
- * values across a barrier.
+ * ## Managing concurrency ##
+ * # Volatile #
+ * A controversial tool in the C toolkit. WiredTiger utilizes volatile for one reason, managing
+ * compiler optimization. Leveraging the definitions from Is Parallel Programming Hard, And, If So,
+ * What Can You Do About It?, Paul E. McKenney, WiredTiger utilizes volatile to prevent the
+ * following type of compiler optimizations: Load fusing, store fusing, invented loads, invented
+ * stores and code reordering. WiredTiger defines the following macros WT_READ_ONCE and
+ * WT_WRITE_ONCE, these macros, if used correctly, prevent the described forms of compiler
+ * optimization. Additionally some variables in WiredTiger are defined with the volatile keyword if
```

Comment:
Can we explain why we need to define the frequently used concurrent variables as volatile?

Reply:
Moving this conversation to: https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1594802348